dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h

```diff
@@ -0,0 +1,211 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_COLOR_MANIPULATION_BRIGHTNESS_CONTRAST_GPU_H_
+#define DALI_KERNELS_IMGPROC_COLOR_MANIPULATION_BRIGHTNESS_CONTRAST_GPU_H_
+
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/core/geom/box.h"
+#include "dali/kernels/common/block_setup.h"
+
+namespace dali {
+namespace kernels {
+namespace brightness_contrast {
+
+
+template <size_t ndims>
+using Roi = Box<ndims, int>;
+
+
+template <class InputType, class OutputType, int ndims>
+struct SampleDescriptor {
+  const InputType *in;
+  OutputType *out;
+  ivec<ndims> in_pitch, out_pitch;
+  float brightness, contrast;
+};
+
+
+template <size_t ndims>
+TensorListShape<ndims> RoiToShape(const std::vector<Roi<ndims>> &rois, int nchannels) {
+  std::vector<TensorShape<ndims>> ret;
+
+  for (auto roi : rois) {
+    assert(all_coords(roi.hi >= roi.lo) && "Cannot create a tensor shape from an invalid Box");
+    TensorShape<ndims> ts;
+    auto e = roi.extent();
+    auto ridx = ndims;
+    ts[--ridx] = e[0] * nchannels;
+    for (size_t idx = 1; idx < ndims; idx++) {
+      ts[--ridx] = e[idx];
+    }
+    ret.emplace_back(ts);
+  }
+
+  return TensorListShape<ndims>{ret};
+}
+
+
+template <size_t ndims>
+std::vector<Roi<ndims>>
+AdjustRois(const std::vector<Roi<ndims>> rois, const TensorListShape<ndims + 1> &shapes) {
+  assert(rois.empty() || rois.size() == static_cast<size_t>(shapes.num_samples()));
+  std::vector<Roi<ndims>> ret(shapes.num_samples());
+
+  auto whole_image = [](auto shape) -> Roi<ndims> {
+      constexpr int spatial_dims = ndims;
+      ivec<spatial_dims> size;
+      for (size_t i = 0; i < spatial_dims; i++)
+        size[i] = shape[spatial_dims - 1 - i];
+      return {0, size};
+  };
+
+  if (rois.empty()) {
+    for (int i = 0; i < shapes.num_samples(); i++) {
+      ret[i] = whole_image(shapes[i]);
+    }
+  } else {
+    for (size_t i = 0; i < rois.size(); i++) {
+      ret[i] = intersection(rois[i], whole_image(shapes[i]));
+    }
+  }
+
+  return ret;
+}
+
+
+template <class OutputType, class InputType, int ndims>
+std::vector<SampleDescriptor<InputType, OutputType, ndims - 1>>
+CreateSampleDescriptors(const InListGPU<InputType, ndims> &in,
+                        const OutListGPU<OutputType, ndims> &out,
+                        const std::vector<float> &brightness, const std::vector<float> &contrast) {
+  std::vector<SampleDescriptor<InputType, OutputType, ndims - 1>> ret(in.num_samples());
+
+  for (int i = 0; i < in.num_samples(); i++) {
+    auto &sample = ret[i];
+    sample.in = in[i].data;
+    sample.out = out[i].data;
+    sample.in_pitch = {};
+    sample.out_pitch = {};
+
+    auto fill_pitch_with_flattening = [](const auto &tv, auto &pitch) {
```

Comment:
this doesn't need to be a lambda. Should be moved to a separate function IMHO.

Other than that, I find the `*--pitch_end *= *--tl_end;` really obscure for what it does

Reply:
```
for (int i = 0; i < pitch.size(); i++) {
    tv.shape[i] = pitch[i]
}
tv.shape[tv.shape.size()-1] *= ...
```